crypto/bls/signature_batch.go

```diff
@@ -33,12 +45,58 @@ func (s *SignatureBatch) Verify() (bool, error) {
 	return VerifyMultipleSignatures(s.Signatures, s.Messages, s.PublicKeys)
 }
 
+// VerifyVerbosely verifies signatures as a whole at first, if fails, fallback
+// to verify each single signature to identify invalid ones.
+func (s *SignatureBatch) VerifyVerbosely() (bool, error) {
+	valid, err := s.Verify()
+	if err != nil || valid {
+		return valid, err
+	}
+
+	// if signature batch is invalid, we then verify signatures one by one.
+
+	var sb strings.Builder
+
+	for i := 0; i < len(s.Signatures); i++ {
+		sig := s.Signatures[i]
+		msg := s.Messages[i]
+		pubKey := s.PublicKeys[i]
+
+		valid, err := VerifySignature(sig, msg, pubKey)
+		if !valid {
+			var desc string
+			if len(s.Descriptions) != len(s.Signatures) {
+				desc = "non-specific signature"
+			} else {
+				desc = s.Descriptions[i]
+			}
+
+			if err != nil {
+				_, e := fmt.Fprintf(&sb, "signature '%s' is invalid."+
+					" signature: %v, public key: %v, message: %v, error: %v\n", desc, sig, pubKey, msg, err)
+				if e != nil {
+					// ignore
+				}
+			} else {
+				_, e := fmt.Fprintf(&sb, "signature '%s' is invalid."+
+					" signature: %v, public key: %v, message: %v\n", desc, sig, pubKey, msg)
```

Comment:
Better to represent `msg` as hex rather than an array of bytes when printing the error message. Same applies to signature and public keys. Also you need to marshal the public key into bytes otherwise you end up printing pointers to the struct which isn't useful

Reply:
Thanks! I'll make error message more readable.